smartsim/_core/launcher/dragon/dragonConnector.py

```diff
@@ -0,0 +1,447 @@
+        self._context.setsockopt(zmq.SNDTIMEO, value=timeout)
+        self._context.setsockopt(zmq.RCVTIMEO, value=timeout)
+        if self._authenticator is not None and self._authenticator.thread is not None:
+            try:
+                self._authenticator.thread.authenticator.zap_socket.setsockopt(
+                    zmq.SNDTIMEO, timeout
+                )
+                self._authenticator.thread.authenticator.zap_socket.setsockopt(
+                    zmq.RCVTIMEO, timeout
+                )
+            except zmq.ZMQError:
+                pass
+
+    def ensure_connected(self) -> None:
+        if not self.is_connected:
+            self.connect_to_dragon()
+        if not self.is_connected:
+            raise SmartSimError("Could not connect to Dragon server")
+
+    def _get_new_authenticator(self) -> None:
+        if self._authenticator is not None:
+            if self._authenticator.thread is not None:
+                try:
+                    logger.debug("Closing ZAP socket")
+                    self._authenticator.thread.authenticator.zap_socket.close()
+                except Exception as e:
+                    logger.debug(f"Could not close ZAP socket, {e}")
+            try:
+                self._authenticator.stop()
+            except zmq.Again:
+                logger.debug("Could not stop authenticator")
+        try:
+            self._authenticator = dragonSockets.get_authenticator(self._context)
+            return
+        except RuntimeError as e:
+            logger.error("Could not get authenticator")
+            raise e from None
+
+    # pylint: disable-next=too-many-statements,too-many-locals
+    def connect_to_dragon(self) -> None:
+        with DRG_LOCK:
+            # TODO use manager instead
+            if self.is_connected:
+                return
+            if self._dragon_server_path is None:
+                raise SmartSimError("Path to Dragon server not set.")
+
+            path = _resolve_dragon_path(self._dragon_server_path)
+            dragon_config_log = path / CONFIG.dragon_log_filename
+
+            if dragon_config_log.is_file():
+
+                dragon_confs = self._parse_launched_dragon_server_info_from_files(
+                    [dragon_config_log]
+                )
+                logger.debug(dragon_confs)
+                for dragon_conf in dragon_confs:
+                    if not "address" in dragon_conf:
+                        continue
+                    logger.debug(
+                        "Found dragon server config file. Checking if the server"
+                        f" is still up at address {dragon_conf['address']}."
+                    )
+                    try:
+                        self._set_timeout(self._timeout)
+                        self._get_new_authenticator()
+                        self._handshake(dragon_conf["address"])
+                    except SmartSimError as e:
+                        logger.debug(e)
```

Comment:
A `SmartSimError` might warrant a call to `logger.error` instead of `logger.debug`

Reply:
Right